### PR TITLE
Clean up ci-signal report and adding cobra

### PR DESCRIPTION
**Pull request by leonardpahlke**
#### What type of PR is this?

/kind cleanup
/kind feature

#### What this PR does / why we need it:

Added features and cleaned up project.
* Added [cobra](https://github.com/spf13/cobra) mentioned in #2314, #2309
* Splitted reports into files (follow cobra conventions)
* Decoupled report data gathering and data representation mentioned in #2314
* Addressed some comments mentioned in #2314

#### Which issue(s) this PR fixes:

None

#### Special notes for your reviewer:

ci signal report is still work in progress. I will focus on one of the points below on the next PR.
* improve performance with go routines (concurrent http calls in particular)
* add tests
* move testgrid and github logic into related packages (&insure there is no duplication)

I think it would make sense to add a cobra command testgrid and github to the report. I added a [note](https://github.com/orgs/kubernetes/projects/65#card-72786227) to the ci-signal project board to discuss this (shouldnt be a lot of work since the structure now has been created). 

#### Does this PR introduce a user-facing change?

ci-signal report out put is the same. Usage is a bit different due to the added main.go file as entrypoint (previously it has been report.go. 

```release-note
The CI-Signal report gets executed from main.go file. Run go run cmd/ci-reporter/main.go to execute the report

Information about commands and flags can be check via the flag -h or --help

Release version flag -v can now also get set via --release-version

Short flag -short now get set via -s or --short

Environment variable GITHUB_AUTH_TOKEN now gets set as GITHUB_TOKEN to align with other projects under k8s
```
/label tide/merge-method-squash
/assign @puerco @palnabarun

### CODE REVIEW

**Comment by leonardpahlke** [comment]
cleaned up my git history on fork

/reopen

**Review by puerco — CHANGES_REQUESTED** [review]
Nice @leonardpahlke! Just a few styling nits. Now that it is split let's try to integrate this as much as we can with the releng libs at k/release-sdk before it starts growing.

File: cmd/ci-reporter/cmd/root.go [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+	PreRun: setGithubConfig,
+	RunE: func(cmd *cobra.Command, args []string) error {
+		return RunReport(*cfg)
+	},
+}
+
+// Execute executes the ci-reporter root command.
+func Execute() error {
+	return rootCmd.Execute()
+}
```
Comment (RinkiyaKeDad) [inline cmd/ci-reporter/cmd/root.go:43]: Does anyone know if we still need the `Execute` function when we don't have subcommands? Wouldn't the 

```
	RunE: func(cmd *cobra.Command, args []string) error {
```

on line 35 be called automatically if there is just one single command?

File: cmd/ci-reporter/cmd/github.go [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+	ColumnTitle string
+	// ColumnID ID of a github project board column
+	ColumnID int64
+)
+
+// GithubProjectBoardColumn specifies a github project board column
+type GithubProjectBoardColumn struct {
+	ColumnTitle ColumnTitle
+	ColumnID    ColumnID
+}
```
Comment (RinkiyaKeDad) [inline cmd/ci-reporter/cmd/github.go:65]: Can be simplified to 

```suggestion
type GithubProjectBoardColumn struct {
	ColumnTitle
	ColumnID   
}
```

File: cmd/ci-reporter/cmd/github.go [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+		}
+	}
+	fmt.Print("\n\n")
+}
+
+func getCardsFromColumn(cfg ReporterConfig, cardsID ColumnID) ([]IssueOverview, error) {
+	opt := &github.ProjectCardListOptions{}
+	cards, _, err := cfg.GithubClient.Projects.ListProjectCards(context.Background(), int64(cardsID), opt)
+	if err != nil {
+		fmt.Printf("error when querying cards %v", err)
```
Comment (RinkiyaKeDad) [inline cmd/ci-reporter/cmd/github.go:119]: Should we use `log.Printf` here instead of `fmt.Printf`?

File: cmd/ci-reporter/cmd/github.go [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+	if err != nil {
+		fmt.Printf("%v", err)
+		return nil, err
+	}
+	var result issueDetail
+	err = json.Unmarshal(body, &result)
+	if err != nil {
+		return nil, err
+	}
+	return &result, nil
```
Comment (RinkiyaKeDad) [inline cmd/ci-reporter/cmd/github.go:181]: Since the struct is relatively simpler and we don't modify it in the `getCardsFromColumn` function, should we return the value instead of a pointer?

File: cmd/ci-reporter/cmd/github.go [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+		}
+		githubReportData[column.ColumnTitle] = cards
+	}
+	return githubReportData, nil
+}
+
+// PrintGithubReportData used to print github report data out to the console
+func PrintGithubReportData(reportData map[ColumnTitle][]IssueOverview) {
+	for columnTitle, issuesInColumn := range reportData {
+		if len(issuesInColumn) > 0 {
```
Comment (RinkiyaKeDad) [inline cmd/ci-reporter/cmd/github.go:105]: If a particular heading has no issues under it I feel we should still print its title for consistency in report outputs, WDYT?

File: cmd/ci-reporter/cmd/testgrid.go [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+		{OutputName: "Master-Blocking", URLName: "sig-release-master-blocking"},
+		{OutputName: "Master-Informing", URLName: "sig-release-master-informing"},
+	}
+
+	if cfg.ReleaseVersion != "" {
+		requiredJobs = append(requiredJobs, []RequiredTestgridJob{
+			{OutputName: cfg.ReleaseVersion + "-blocking", URLName: "sig-release-" + cfg.ReleaseVersion + "-blocking"},
+			{OutputName: cfg.ReleaseVersion + "-informing", URLName: "sig-release-" + cfg.ReleaseVersion + "-informing"},
+		}...)
+	}
```
Comment (RinkiyaKeDad) [inline cmd/ci-reporter/cmd/testgrid.go:44]: I think all of this can be simplified by setting `cfg.ReleaseVersion` to a default value of `master` and then doing:

```suggestion
		requiredJobs = append(requiredJobs, []RequiredTestgridJob{
			{OutputName: cfg.ReleaseVersion + "-blocking", URLName: "sig-release-" + cfg.ReleaseVersion + "-blocking"},
			{OutputName: cfg.ReleaseVersion + "-informing", URLName: "sig-release-" + cfg.ReleaseVersion + "-informing"},
		}...)
```

File: cmd/ci-reporter/cmd/testgrid.go [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+
+// GetTestgridReportData used to request the raw report data from testgrid
+func GetTestgridReportData(cfg ReporterConfig) ([]TestgridStatistics, error) {
+	requiredJobs := []RequiredTestgridJob{
+		{OutputName: "Master-Blocking", URLName: "sig-release-master-blocking"},
+		{OutputName: "Master-Informing", URLName: "sig-release-master-informing"},
+	}
+
+	if cfg.ReleaseVersion != "" {
+		requiredJobs = append(requiredJobs, []RequiredTestgridJob{
```
Comment (RinkiyaKeDad) [inline cmd/ci-reporter/cmd/testgrid.go:40]: Also if a release version has been specified why are we appending instead of replacing? Do we still want to show the report for master in that case?

File: cmd/ci-reporter/cmd/testgrid.go [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+	Total   int
+	Passing int
+	Flaking int
+	Failing int
+	Stale   int
+}
+
+type overview struct {
+	OverallStatus string `json:"overall_status"`
+}
```
Comment (RinkiyaKeDad) [inline cmd/ci-reporter/cmd/testgrid.go:113]: Let's define the types before the functions for better readability 🙂

**Review by RinkiyaKeDad — CHANGES_REQUESTED** [review]
Thanks a lot for this @leonardpahlke! ❤️ A couple of changes needed before we can land this 🙂 Can you also please add tests for this code since right now we've no way of testing all this? Thanks again!

/hold for changes and tests

File: cmd/ci-reporter/README.md [hunk 9]
```diff
@@ -3,7 +3,7 @@
 You can get the current overview for CI signal report by running
 
 ```
-GITHUB_AUTH_TOKEN=xxx go run cmd/ci-reporter/report.go
+GITHUB_AUTH_TOKEN=xxx go run cmd/ci-reporter/main.go
```
[the remark this answers is not in the document]
Comment (leonardpahlke) [reply cmd/ci-reporter/README.md]: > go get [URL] to grab the binary for the CLI and then run reporter

I think we should do that after we clarified if we like to merge it with krel or not. Also we should wait until we have made all major changes to the code which we plan todo.

File: cmd/ci-reporter/main.go [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+import (
+	"os"
+
+	"github.com/sirupsen/logrus"
+	"k8s.io/release/cmd/ci-reporter/cmd"
+)
+
+func main() {
+	if err := cmd.Execute(); err != nil {
+		logrus.Fatal(err)
```
[the remark this answers is not in the document]
Comment (leonardpahlke) [reply cmd/ci-reporter/main.go:28]: Its used in this project (otherwise you would see a new import in go.mod :wink:)

File: cmd/ci-reporter/cmd/github.go (hunk 4 above)
Comment (leonardpahlke) [reply cmd/ci-reporter/cmd/github.go:179]: Thank you for that, this relates to the version that alena provided since this is part of the github reporter I will come back to that in one of the next PRs and check out which approach we can take.

File: cmd/ci-reporter/cmd/github.go (hunk 5 above)
Comment (leonardpahlke) [reply cmd/ci-reporter/cmd/github.go:105]: We likely change the format of the report in the future. (I was experimenting with a generic reporter form which abstracts the testrgid and github reporter, this may be the way to go - but not in this PR)

File: cmd/ci-reporter/cmd/testgrid.go (hunk 6 above)
Comment (leonardpahlke) [reply cmd/ci-reporter/cmd/testgrid.go:44]: Since we always like to see master blocking / master informing we can not further simplify this. But thanks for the thought - we can discuss if this would be a better way to generate testgrid reports

File: cmd/ci-reporter/cmd/testgrid.go (hunk 8 above)
Comment (leonardpahlke) [reply cmd/ci-reporter/cmd/testgrid.go:113]: I will come back to that in one of the nexr PRs (testgrid align with existing pkgs)

**Comment by leonardpahlke** [comment]
@puerco thank you for your comments. Code has been updated. Next PR will focus on testgrid / github.

@RinkiyaKeDad thank you for writeing all these comments. I wrote answers to your comments.

> Can you also please add tests for this code since right now we've no way of testing all this? Thanks again!

planned (as you can see in the PR description)

File: cmd/ci-reporter/cmd/root.go (hunk 1 above)
Comment (leonardpahlke) [reply cmd/ci-reporter/cmd/root.go:43]: cobra convention

File: cmd/ci-reporter/cmd/github.go (hunk 2 above)
Comment (leonardpahlke) [reply cmd/ci-reporter/cmd/github.go:65]: I like it more explicit, so current version (i think it easier to understand also for someone not so familiar with go)

File: cmd/ci-reporter/cmd/github.go (hunk 3 above)
Comment (leonardpahlke) [reply cmd/ci-reporter/cmd/github.go]: resolved (puerco pointed error return statements out too)

File: cmd/ci-reporter/cmd/testgrid.go (hunk 7 above)
Comment (leonardpahlke) [reply cmd/ci-reporter/cmd/testgrid.go:40]: same as comment above

**Comment by RinkiyaKeDad** [comment]
@leonardpahlke I'm a bit hesitant about merging this code without having tests for it in the same PR. Having tests in a separate PR would make working on the tool for other contributors tough because we won't be able to identify breaking changes till the PR with the tests lands. Could you please add tests in this PR itself? 🙏  Let me know if you need any help with those!

**Comment by leonardpahlke** [comment]
@RinkiyaKeDad adding tests to the code right now does not make sense. We need to put the logic into the related packages first. As already said, that part is next on the list.

**Comment by puerco** [comment]
In general, I am also +1 on having the tests merge with the new functions. The process of this tool has been a bit different because we are moving and refactoring an external project but in general we try to do it that way. @leonardpahlke what is blocking you from sending the tests in the same PR?

**Comment by leonardpahlke** [comment]
So, nothing is blocking me from further enhancing the code at this point. From what we have discussed before, I understood that small PR's would be preferred to simplify reviews and dont overload one comment with a lot of changes. 

This PR does some of the neccessary changes we need to do before integrating any new features ([which currently under discussion](https://github.com/orgs/kubernetes/projects/65)). These changes are from nature *obvious* or *alignment reasons* regarding integrating it into k/release.

Things todo before adding new features (as already wrote in previous PRs):
1. general refactoring (done in previous PR)
2. add cobra (*align with k/release*)
3. split reporter code into files
4. align testgrid pkg with testgrid reporter methods (*align with k/release*) + **add tests**
5. align github pkg with github reporter methods (*align with k/release*) + **add tests**
6. add go routines & improve performance

The second and third point have been assessed with this PR. If I add tests to the PR without checking 4 & 5 this would be a waste of time as the code will change next PR anyways.
If I would add tests to the PR & also check 4 & 5 the PR would be quite big.

Both options are fine for me but I already asked if this approach (spreading these necessary changes out into multiple PR's) would be fine and now @RinkiyaKeDad is questioning it again - which is quite frustrating tbh as this is a huge waste of time to explain this.

Long story short, @puerco I can add tests to the code. If so, should I also check number 3 & 4 of the list or not? Thank you for taking the time.

**Comment by RinkiyaKeDad** [comment]
> and now @RinkiyaKeDad is questioning it again - which is quite frustrating tbh as this is a huge waste of time to explain this.

@leonardpahlke I'm really sorry if my reviews frustrated you or wasted your time. Adding tests in the same PR was just a suggestion from my side but if you feel that's not the way to go I don't want to be a blocker. 

/unhold from my side

**Comment by puerco** [comment]
Yes, now we have molded these functions and values a little bit more like the rest of the codebase. As I was saying, in the case of this tool we are going through a refactor to adapt the donated code to the rest of the tools here. If these functions are just in a transitional state, I'm happy if you add the tests in the final packages.

**Comment by leonardpahlke** [comment]
Okay, great! So we are ready to merge this PR :) After that I will continue with the list of todo's I mentioned before.

Thank you all for your time reviewing this PR!!

**Comment by puerco** [comment]
Sounds good, looking forward to the follow-ups @leonardpahlke 
/lgtm
/approve
/hold cancel
